### PR TITLE
Core dev guidelines

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
### Core Development Guidelines

This is a long overdue write-up on the proposed [core development guidelines](https://github.com/fair-acc/graph-prototype/blob/coreDevGuidelines/CORE_DEVELOPMENT_GUIDELINE.md) primarily targeted at maintainers, users that want to contribute to the core BlockLib, new Schedulers, or new features to the GNU Radio Core.

**Brief Summary:** *(please consult the [full text](https://github.com/fair-acc/graph-prototype/blob/coreDevGuidelines/CORE_DEVELOPMENT_GUIDELINE.md) and references for details)*

The Core Development Guidelines are centred around creating sustainable, efficient, and adaptable software, specifically for GNU Radio.

### Core Goals
1. **Fit-for-Purpose Software**: The software should be adaptable to changing requirements, particularly for SDR applications.
2. **Friendly and Empathetic Development**: Developers should be considerate and helpful towards others.
3. **Lean and Clean Software**: Follow best practices to write efficient and maintainable software.

### Target Audience and Application Classes
The guidelines categorize developers into four application classes, each with different coding rigor:
- **Class 0**: Private users tinkering with GNU Radio.
- **Class 1**: Developers of Out-Of-Tree Modules, Blocks, or Schedulers.
- **Class 2**: Contributors to GNU Radio Core BlockLib with long-term support.
- **Class 3**: Developers of new features in the GNU Radio Core.

### Fundamental Principles
- **Lean and Clean Code**: Emphasizes quantifiable metrics, waste elimination, and clear, maintainable code.
- **W.A.R.M.E.D Approach**: A mnemonic for Write, Agree, Readable, Modify, Execute with Efficiency, and Debug Strategically. Each principle has a metric for optimization.

### Specific Development Guidelines
1. **Common Sense and Professional Conduct**: Adherence to the Code of Conduct and Contribution Guidelines.
2. **Best Practices**: Follow C++ Core Guidelines, continuous learning, and improvement in coding.
3. **Tool Mastery**: Use of linters, static code analysis tools, and automated code formatting.
4. **Coding Best Practices**: Includes minimizing variable scope, unique naming, and concise functions.
5. **C++ Standards**: Use the latest supported C++ standards and STL.
6. **Minimize 3rd-Party Dependencies**: Only crucial or agreed-upon libraries should be used.
7. **Effective Testing and Documentation**: Ensure functional unit testing and meaningful code coverage.
8. **Naming Convention**: Follow specific naming guidelines.
9. **Structs vs Classes**: Prefer functional programming and public structs over classes.
10. **Header-only Code**: Favor header-only code for optimization.
11. **Contemporary C++ Practices**: Appropriate use of modern C++ features and practices.

### Resources and Learning
The document also provides resources for further learning and development in C++ and software engineering practices.

### Appendix
Details specific recommendations for each application class, emphasizing the importance of adhering to guidelines and best practices relevant to the level of contribution to the GNU Radio project.

---

🌟 **Hey GNU Radio Devs & Enthusiasts!** 🌟

Before we dive into the world of code and creativity, here's a quick reminder to set the stage for a positive and productive environment:

1. **Keep It Cool and Constructive**: We're all here to learn, grow, and contribute to GNU Radio. Let's keep our discussions civil, enjoyable, and professional. Think of this as a coffee chat with fellow enthusiasts – relaxed yet respectful.

2. **Respect the Code-of-Conduct**: Our [Code-of-Conduct](https://github.com/fair-acc/graph-prototype/blob/coreDevGuidelines/CODE_OF_CONDUCT.md) is our guide to ensure everyone feels welcome and valued. Let's honour it in spirit and letter.

3. **Assume Good Will**: We're a team, each with the best intentions. Misunderstandings happen, but let's approach each conversation with positivity and the benefit of the doubt.

4. **Value Opinions, Love Data**: Your opinions and ideas are vital, but backing them up with quantifiable and objective metrics (like our W.A.R.M.E.D. approach) makes them even more impactful.

5. **Avoid the Bike-Shed**: Let's focus on the big picture in our discussions to avoid getting lost in minor details. Think 'forest', not just 'trees'.

Let's collaborate in a way that makes us all proud and create something awesome together! 🚀

### CODE REVIEW

File: CONTRIBUTING.md
Comment (Notou) [inline CONTRIBUTING.md:1]: It doesn't seem to be directly related to the commit at hand,
but why is there mention of Java stuff in there?
Lines 31 and 32, it talks about JDK11 and JUnit.

Is that an artifact of copypasting from another repo?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#Contributing to OpenCMW
+#Contributing to GNU Radio
```
Comment (drslebedev) [inline CONTRIBUTING.md:1]: Missing space

```suggestion
# Contributing to GNU Radio
```

File: CONTRIBUTING.md
Comment (RalphSteinhagen) [reply CONTRIBUTING.md:1]: yes, was an artifact from another twinned project of us. Thanks for spotting this.

**Review by mormj — APPROVED** [review]
TODO: We need to clearly from the README direct people here that are "Core Contributors" - i.e. touching the guts of the code.  I think that it is laid out in the application classes, but we don't want to scare potential contributors off by a daunting set of guidelines.  IMO this is ok to merge, but before 4.0 is released there needs to be a documentation effort around directing the classes of users to the right place.

**Pull request merged**